### PR TITLE
First pass at making blocking I/O calls in thread pool.

### PR DESCRIPTION
**Pull request by ry**
This doesn't attempt to convert all of the handlers, just tries to establish the pattern.

### CODE REVIEW

File: src/handlers.rs [hunk 1]
```diff
@@ -440,56 +442,85 @@ fn handle_make_temp_dir(state: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
...
+
+// TODO Do not use macro for the blocking function.. We should instead be able
+// to do this with a normal function, but there seems to some type system
+// issues. The type of this function should be something like this:
+//   fn blocking<F>(is_sync: bool, f: F) -> Box<Op>
+//   where F: FnOnce() -> DenoResult<Buf>
+macro_rules! blocking {
+  ($is_sync:expr,$fn:expr) => {
+    if $is_sync {
+      // If synchronous, execute the function immediately on the main thread.
```
Comment (piscisaureus) [inline src/handlers.rs]: The issue with this is, the blocking function can never have a dependency on an asynchronous function.

Random example:

```
writeFile("somefile", "hello"); <-- async
readFileSync("somefile"); <-- this should wait for writeFile to finish.
```

Even if you don't agree with the above semantics, I predict there will be cases where you do agree.
So I think this is not the right direction -- it'd be better to do block_on(future) instead.

File: src/handlers.rs [hunk 2]
```diff
@@ -440,56 +442,85 @@ fn handle_make_temp_dir(state: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
...
+    Err(_) => panic!("blocking error"),
+  }
+}
+
+// TODO Do not use macro for the blocking function.. We should instead be able
+// to do this with a normal function, but there seems to some type system
+// issues. The type of this function should be something like this:
+//   fn blocking<F>(is_sync: bool, f: F) -> Box<Op>
+//   where F: FnOnce() -> DenoResult<Buf>
+macro_rules! blocking {
```
Comment (piscisaureus) [inline src/handlers.rs]: Using a macro for this is a great idea.

File: src/handlers.rs [hunk 3]
```diff
@@ -440,56 +442,85 @@ fn handle_make_temp_dir(state: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
...
 fn handle_read_file(_config: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
   let msg = base.msg_as_read_file().unwrap();
   let cmd_id = base.cmd_id();
-  let filename = String::from(msg.filename().unwrap());
-  Box::new(futures::future::result(|| -> OpResult {
-    debug!("handle_read_file {}", filename);
-    let vec = fs::read(Path::new(&filename))?;
+  let filename = PathBuf::from(msg.filename().unwrap());
+  debug!("handle_read_file {}", filename.display());
+  blocking!(base.sync(), || {
```
Comment (piscisaureus) [inline src/handlers.rs]: I like that it is this easy to implement both sync and async versions of the same function.

File: src/handlers.rs (hunk 1 above)
Comment (ry) [reply src/handlers.rs]: The idea with the sync fs functions is that we want to enable people to issue simple syscalls and bipass the complex event machine. Sometimes you want to just block everything and run "mkdir". What your suggesting essentially makes the sync calls into a new form of "await" (with questionable semantics), which blocks the event loop. Now sync calls are no longer syscall invocations - they are complex beasts.

I understand this "chaining" idea you're thinking about, but this isn't the place for it. The sync calls are to bipass all complexity and simply execute a syscall immediately on the main thread.

File: src/handlers.rs (hunk 2 above)
Comment (piscisaureus) [reply src/handlers.rs]: Rubberstamp bot rubberstamp.

**Pull request merged**